// fast-glob/out/utils/pattern.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.removeDuplicateSlashes = exports.matchAny = exports.convertPatternsToRe = exports.makeRe = exports.getPatternParts = exports.expandBraceExpansion = exports.expandPatternsWithBraceExpansion = exports.isAffectDepthOfReadingPattern = exports.endsWithSlashGlobStar = exports.hasGlobStar = exports.getBaseDirectory = exports.isPatternRelatedToParentDirectory = exports.getPatternsOutsideCurrentDirectory = exports.getPatternsInsideCurrentDirectory = exports.getPositivePatterns = exports.getNegativePatterns = exports.isPositivePattern = exports.isNegativePattern = exports.convertToNegativePattern = exports.convertToPositivePattern = exports.isDynamicPattern = exports.isStaticPattern = void 0;
const path = require("path");
const globParent = require("glob-parent");
const micromatch = require("micromatch");
const GLOBSTAR = '**';
const ESCAPE_SYMBOL = '\\';
const COMMON_GLOB_SYMBOLS_RE = /[*?]|^!/;
const REGEX_CHARACTER_CLASS_SYMBOLS_RE = /\[[^[]*]/;
const REGEX_GROUP_SYMBOLS_RE = /(?:^|[^!*+?@])\([^(]*\|[^|]*\)/;
const GLOB_EXTENSION_SYMBOLS_RE = /[!*+?@]\([^(]*\)/;
const BRACE_EXPANSION_SEPARATORS_RE = /,|\.\./;
/**
 * Matches a sequence of two or more consecutive slashes, excluding the first two slashes at the beginning of the string.
 * The latter is due to the presence of the device path at the beginning of the UNC path.
 */
const DOUBLE_SLASH_RE = /(?!^)\/{2,}/g;
function isStaticPattern(pattern, options = {}) {
    return !isDynamicPattern(pattern, options);
}
exports.isStaticPattern = isStaticPattern;
function isDynamicPattern(pattern, options = {}) {
    /**
     * A special case with an empty string is necessary for matching patterns that start with a forward slash.
     * An empty string cannot be a dynamic pattern.
     * For example, the pattern `/lib/*` will be spread into parts: '', 'lib', '*'.
     */
    if (pattern === '') {
        return false;
    }
    /**
     * When the `caseSensitiveMatch` option is disabled, all patterns must be marked as dynamic, because we cannot check
     * filepath directly (without read directory).
     */
    if (options.caseSensitiveMatch === false || pattern.includes(ESCAPE_SYMBOL)) {
        return true;
    }
    if (COMMON_GLOB_SYMBOLS_RE.test(pattern) || REGEX_CHARACTER_CLASS_SYMBOLS_RE.test(pattern) || REGEX_GROUP_SYMBOLS_RE.test(pattern)) {
        return true;
    }
    if (options.extglob !== false && GLOB_EXTENSION_SYMBOLS_RE.test(pattern)) {
        return true;
    }
    if (options.braceExpansion !== false && hasBraceExpansion(pattern)) {
        return true;
    }
    return false;
}
exports.isDynamicPattern = isDynamicPattern;
function hasBraceExpansion(pattern) {
    const openingBraceIndex = pattern.indexOf('{');
    if (openingBraceIndex === -1) {
        return false;
    }
    const closingBraceIndex = pattern.indexOf('}', openingBraceIndex + 1);
    if (closingBraceIndex === -1) {
        return false;
    }
    const braceContent = pattern.slice(openingBraceIndex, closingBraceIndex);
    return BRACE_EXPANSION_SEPARATORS_RE.test(braceContent);
}
function convertToPositivePattern(pattern) {
    return isNegativePattern(pattern) ? pattern.slice(1) : pattern;
}
exports.convertToPositivePattern = convertToPositivePattern;
function convertToNegativePattern(pattern) {
    return '!' + pattern;
}
exports.convertToNegativePattern = convertToNegativePattern;
function isNegativePattern(pattern) {
    return pattern.startsWith('!') && pattern[1] !== '(';
}
exports.isNegativePattern = isNegativePattern;
function isPositivePattern(pattern) {
    return !isNegativePattern(pattern);
}
exports.isPositivePattern = isPositivePattern;
function getNegativePatterns(patterns) {
    return patterns.filter(isNegativePattern);
}
exports.getNegativePatterns = getNegativePatterns;
function getPositivePatterns(patterns) {
    return patterns.filter(isPositivePattern);
}
exports.getPositivePatterns = getPositivePatterns;
/**
 * Returns patterns that can be applied inside the current directory.
 *
 * @exampДуртай","Дургүй","Дэлгэцийн зургийг оруул","Дотоод Microsoft","Нууцлалын бодлого","Илгээх","{0} баг руу э-шуудан илгээх",,"Нэмж мэдэх","Нууцлалын баталгаа","Асуудлыг мэдээлэх","Windows хайлт дээр өөрийн санал хүсэлтээ оруулна уу. Таны хувийн нууцлалыг хамгаалахын тулд хувийн мэдээллийг бүү оруул.","Microsoft руу Санал хүсэлт илгээх"])("mr",["कृपया आपले टोपणनाव प्रविष्ट करा","आणि मला येथे cc करा","आपले उपाख्य येथे भरा.","एखादा विशिष्ट फीडबॅक मिळाला का?","रद्द करा","कृपया अभिप्राय द्या.","आपल्या अभिप्रायाचा मजकूर लिहीण्यासाठी जागा.","फीडबॅकचा प्रकार","सुचवा","लाइक","डिस्लाइक करा","कल्पना","लाइक्स","डिस्लाइक करतात","स्क्रीनशॉट समाविष्ट करा","Microsoft इंटरनल","गोपनीयता धोरण","पाठवा","{0} टीमला इमेल पाठवा","हे बटण क्लिक केल्याने किंवा enter दाबल्याने फीडबॅक यशस्वीरित्या सबमिट केला जाईल","अधिक जाणून घ्या","प्रायव्हसी स्टेटमेंट","चिंतेची नोंद करा","येथे Windows Search संबंधित आपला फीडबॅक प्रविष्ट करा. आपल्या गोपनियतेच्या संरक्षणेसाठी, कृपया कोणतीही व्यक्तिगत माहिती समाविष्ट करू नका.","Microsoft ला फीडबॅक पाठवा"])("ms",["Sila masukkan alias anda.","dan sk saya di","Masukkan alias anda di sini.","Dapat sebarang maklum balas terperinci?","Batal","Sila tinggalkan komen.","Kotak teks untuk maklum balas anda","Jenis maklum balas","Cadang","Suka","Tidak suka","Idea","Suka","Tidak suka","Termasuk paparan skrin","Dalaman Microsoft","Dasar privasi","Hantar","Hantar e-mel kepada pasukan {0}","Mengklik butang ini atau menekan Enter akan berjaya menyerahkan maklum balas","Ketahui selanjutnya","Kenyataan Privasi","Laporkan kebimbangan","Masukkan maklum balas anda dalam Window Search di sini. Untuk melindungi privasi anda, jangan masukkan maklumat peribadi.","Hantarkan maklum balas kepada Microsoft"])("mt",["Jekk jogħġbok daħħal l-alias tiegħek.","u cc lili fuq","Daħħal l-alias tiegħek hawnhekk.","Għandek xi rispons speċifiku?","Ikkanċella","Jekk jogħġbok ħalli kumment.","Kaxxa tat-test għall-feedback tiegħek","Tip ta' feedback","Issuġġerixxi","Jogħġobni","Ma jogħġobnix","Ideat","Ma jogħġbunix","Ma jogħġbunix","Inkludi ritratt tal-iskrin",,"Politika tal-privatezza","Ibgħat","Ibgħat email lit-tim {0}","Jekk tikklikkja din il-buttuna jew tagħfas enter se tissottometti l-feedback b’suċċess","Sir af aktar","Dikjarazzjoni tal-Privatezza","Irrapporta xi tħassib","Daħħal il-feedback tiegħek fuq Windows Search hawnhekk. Biex tipproteġi l-privatezza tiegħek, jekk jogħġbok tinkludi ebda informazzjoni personali.","Ibgħat feedback lil Microsoft"])("nb-no",["Angi aliaset ditt.","og kopi til meg på","Skriv inn aliaset ditt her.","Har du noen spesielle tilbakemeldinger?","Avbryt","Legg igjen en kommentar",
